Fix payload and defaults

twangboy opened the pull request:
### What does this PR do?
Adds a new Constant `NOT_SET` to allow `None` to be passed as a valid option to bypass throwing an error. We can't set default to be an object because msgpack can't serialize it and causes `sys.argparse` to fail when returning data to the master.

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/62450

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes

twangboy committed:
Allow default to be none

Ch3LL reviewed (changes requested):
changelog? and does there need to be test coverage?

twangboy commented:
There was a test written in the original PR. I think it covers the changes here:
https://github.com/saltstack/salt/pull/61089

Ch3LL commented:
It doesn't look like those tests are covering the issue created in https://github.com/saltstack/salt/issues/62450 . It should be easy to add a test that ensures none is returned when default none is passed. 

`salt-call --local pillar.get doesnotexist default=None`

twangboy committed:
Additional tests

frebib commented:
I tend to agree with Wayne's comment (https://github.com/saltstack/salt/issues/62450#issuecomment-1212262538) suggesting to use a static constant instead of magic behaviour based on internal python values to avoid any chance of conflict or ambiguity. Arguably using `None` in the first place is the same idea as using `...` but just with a more commonly used constant

s0undt3ch commented:
> I tend to agree with Wayne's comment ([#62450 (comment)](https://github.com/saltstack/salt/issues/62450#issuecomment-1212262538)) suggesting to use a static constant instead of magic behaviour based on internal python values to avoid any chance of conflict or ambiguity. Arguably using `None` in the first place is the same idea as using `...` but just with a more commonly used constant

A static constant would involve more code to support it(more imports in the code base).

`...` is not magical, [it's a built-in constant](https://docs.python.org/3/library/constants.html#Ellipsis), and in this context, it serves the exact same purpose as `None` without the side-effects that `None` caused, and with no additional imports in the code.

s0undt3ch commented:
Would this cause less potential confusion if we used `False` instead of `...`?

s0undt3ch committed:
Switch away from using Elipsis

```
Python 3.7.12 (default, Mar 16 2022, 11:48:18)
[GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pickle
>>> import salt.defaults
>>> salt.defaults.NOT_SET
<NOT_SET>
>>> d = pickle.dumps(salt.defaults.NOT_SET)
>>> ld = pickle.loads(d)
>>> ld
<NOT_SET>
>>> ld == salt.defaults.NOT_SET
True
>>>
```

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch commented:
After some internal discussion, I retracted my proposal and we implemented a "constant".

waynew commented on salt/defaults/__init__.py:
```diff
@@ -5,5 +5,58 @@
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return False
+        if self.name != other.name:
+            return False
+        return self.value == other.value
```
**suggestion (non-blocking)** could be `return self.name == other.name and self.value == other.value`, but that's not an essential change

waynew reviewed (approved):
One thought, but it's not a blocker at all. This looks good - I didn't run the tests locally, so trusting that happened correctly on your end :slightly_smiling_face:
